optics/src/main/scala/io/circe/optics/JsonPath.scala

```diff
@@ -0,0 +1,27 @@
+package io.circe.optics
+
+import io.circe.{ Decoder, Encoder, Json }
+import io.circe.optics.JsonOptics._, io.circe.optics.JsonObjectOptics._
+import monocle.{ Optional, Prism }
+import monocle.function.index
+import monocle.std.list._
+import scala.language.dynamics
+
+case class JsonPath(opt: Optional[Json, Json]) extends Dynamic {
+  def selectDynamic(field: String): JsonPath =
+    JsonPath(opt.composePrism(jsonObject).composeOptional(index(field)))
+
+  def at(i: Int): JsonPath =
+    JsonPath(opt.composePrism(jsonArray).composeOptional(index(i)))
+
+  def as[A](implicit decode: Decoder[A], encode: Encoder[A]): Optional[Json, A] =
```

Comment:
I believe that `as` is not lawful because `decoder` is not injective e.g.
```scala
case class Point(x: Int, y: Int)

val decoder: Decoder[Point] = ...
decoder.decodeJson(obj("x" -> int(1), "y" -> int(2))) == decoder.decodeJson(obj("x" -> int(1), "y" -> int(2), "z" -> int(5))) 
```